bayes_opt/bayesian_optimization.py

```diff
@@ -260,21 +283,146 @@ def maximize(self,
+            self.save_res_to_csv(True, exact_x=exact_x_dict)
+
+        else:
+            while not self._queue.empty or iteration < n_iter:
+                try:
+                    x_probe = next(self._queue)
+                except StopIteration:
+                    util.update_params()
+                    x_probe = self.suggest(util)
+                    iteration += 1
+
+                self.probe(x_probe, lazy=False)
+
+                if self._bounds_transformer:
+                    self.set_bounds(
+                        self._bounds_transformer.transform(self._space))
+
+            self.dispatch(Events.OPTIMIZATION_END)
+            self.save_res_to_csv(False)
+
+    def get_approximation(self, dataset, x_probe):
+        """
+        Method to get from the dataset passed by the user the nearest point to the x_probe point
+
+        Parameters
+        ----------
+
+        dataset: pandas.DataFrame
+            dataset specified by the user
+
+        x_probe: dict
+            point found by the optimization process
+        """
+
+        try:
+            x_array = numpy.array(list(x_probe.values()))
+
+        except AttributeError:
+            x_array = x_probe
+
+        min_distance = None
+        min_index = None
+        approximations = []
+
+        if self._target_column is None:
+            for row in dataset.itertuples():
+                dataset_tuple = []
+                for j in range(x_array.size):
+                    dataset_tuple.append(row[j + 1])
+
+                res = numpy.linalg.norm(x_array - dataset_tuple, 2)
+
+                if min_distance is None:
+                    min_distance = res
+                    approximations = self._space.array_to_params(dataset_tuple)
+                elif res == min_distance:
+                    approximations.append(self._space.array_to_params(dataset_tuple))
+                elif res < min_distance:
+                    min_distance = res
+                    approximations = self._space.array_to_params(dataset_tuple)
+            return random.choice(approximations)
+        else:
+            # dataset.drop(self._target_column)
+            for row in dataset.loc[:, dataset.columns != self._target_column].itertuples():
+                dataset_tuple = []
+                for j in range(x_array.size):
+                    dataset_tuple.append(row[j + 1])
+
+                res = numpy.linalg.norm(x_array - dataset_tuple, 2)
+
+                if min_distance.size == 0:
+                    min_index = row[0]
+                    min_distance = res
+                    approximations = self._space.array_to_params(dataset_tuple)
+                elif res == min_distance:
+                    if row[self._target_column] > dataset.iloc[min_index][self._target_column]:
+                        min_index = row[0]
+                        min_distance = res
+                        approximations = self._space.array_to_params(dataset_tuple)
```

Comment:
It looks like you're recovering the database entry whose value in the target column is the largest. This is not what should happen. The target column value has no influence on the choice of the best approximation. If you have multiple entries with minimal distance from `x_probe`, you pick one at random, just like you did in the case without `target_column`. Remember that we're trying to simulate the actual execution of a program: sometimes it runs for a shorter time, sometimes longer

Reply:
now the row is randomly chosen